stimpacks/deploy/deploy.go

```diff
@@ -0,0 +1,102 @@
+package deploy
+
+import (
+	"strings"
+
+	log "github.com/PremiereGlobal/stim/pkg/stimlog"
+	"github.com/PremiereGlobal/stim/stim"
+)
+
+const (
+	allOptionPrompt = "--ALL--"
+	allOptionCli    = "all"
+)
+
+// Deploy is the primary type for the stim deploy subcommand
+type Deploy struct {
+	name   string
+	stim   *stim.Stim
+	config Config
+	log    log.StimLogger
+}
+
+// New creates a new 'Deploy' object
+func New() *Deploy {
+	deploy := &Deploy{}
+	return deploy
+}
+
+// Name is a required stim function that returns the name of the stimpack
+func (d *Deploy) Name() string {
+	return d.name
+}
+
+// Run is the main entrypoint to the "deploy" command
+func (d *Deploy) Run() {
+
+	d.log = d.stim.GetLogger()
+
+	// Read in the config file and set up defaults
+	d.parseConfig()
+
+	// Determine the selected environment (via cli param) or prompt the user
+	selectedEnvironmentName := ""
+	environmentArg := d.stim.GetConfig("deploy.environment")
+	if environmentArg != "" {
+		if _, ok := d.config.environmentMap[environmentArg]; ok {
+			selectedEnvironmentName = environmentArg
+		} else {
+			d.log.Fatal("Provided environment value '{}' is not in config file", environmentArg)
+		}
+	} else {
+		environmentList := make([]string, len(d.config.Environments))
+		for i, e := range d.config.Environments {
+			environmentList[i] = e.Name
+		}
+		selectedEnvironmentName, _ = d.stim.PromptList("Which environment?", environmentList, d.stim.GetConfig("deploy.environment"))
+	}
+	selectedEnvironment := d.config.Environments[d.config.environmentMap[selectedEnvironmentName]]
+
+	// Determine the selected instance (via cli param) or prompt the user
+	selectedInstanceName := ""
+	instanceArg := d.stim.GetConfig("deploy.instance")
+	if strings.ToLower(instanceArg) == strings.ToLower(allOptionCli) {
+		selectedInstanceName = allOptionCli
+	} else if instanceArg != "" {
+		if _, ok := selectedEnvironment.instanceMap[instanceArg]; ok {
+			selectedInstanceName = instanceArg
+		} else {
+			d.log.Fatal("Provided instance value '{}' is not in config file under environment '{}'", instanceArg, selectedEnvironmentName)
+		}
+	} else {
+		instanceList := make([]string, len(selectedEnvironment.Instances)+1)
+		instanceList[0] = allOptionPrompt
+		for i, inst := range selectedEnvironment.Instances {
+			instanceList[i+1] = inst.Name
+		}
+
+		selectedInstanceName, _ = d.stim.PromptList("Which instance?", instanceList, d.stim.GetConfig("deploy.instance"))
+	}
```

Comment:
a possible better format for this section would look like this:
```golang
instanceList := make([]string, len(selectedEnvironment.Instances)+1)
copy(instanceList[1:], selectedEnvironment.Instances)
instanceList[0] = allOptionPrompt
selectedInstanceName, _ := d.stim.PromptList("Which instance?", instanceList, d.stim.GetConfig("deploy.instance"))
if strings.ToLower(selectedInstanceName) == strings.ToLower(allOptionCli) {
	selectedInstanceName = allOptionCli
} else {
	if _, ok := selectedEnvironment.instanceMap[instanceArg]; ok {
		selectedInstanceName = instanceArg
	} else {
		d.log.Fatal("Provided instance value '{}' is not in config file under environment '{}'", instanceArg, selectedEnvironmentName)
	}
} 
```

A simalar pattern could be used for line 43 for the `selectedEnvironmentName`